roc.go

```diff
@@ -0,0 +1,145 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import "sort"
+
+// ROC returns paired false positive rate (FPR) and true
+// positive rate (TPR) values corresponding to n
+// cutoffs spanning the relative (or receiver)
+// operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for classes with weights.
+//
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+//
+// The input y must be sorted, and SortWeightedLabeled can be used
+// in order to sort y together with classes and weights.
+//
+// For a given cutoff value, observations corresponding to
+// entries in y greater than the cutoff value are classified as
+// false, while those below (or equal to) the cutoff value are classified
+// as true. These assigned class labels are compared with the true values
+// in the classes slice and used to calculate the FPR and TPR.
+//
+// If weights is nil, all weights are treated as 1.
+//
+// n of zero results in all possible cutoffs being calculated.
+//
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []float64) {
+	if len(y) != len(classes) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(y) != len(weights) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.Float64sAreSorted(y) {
+		panic("stat: input must be sorted")
+	}
+
+	var incWidth, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot calculate fewer than 2 points on a ROC curve")
+		}
+		incWidth = (y[len(y)-1] - y[0]) / float64(n-1)
+		if incWidth <= 0 {
+			panic("stat: input constant")
+		}
+		tol = y[0] + incWidth
+	}
+
+	if n > (len(y) + 1) {
+		tpr = make([]float64, n)
+		fpr = make([]float64, n)
+	} else {
+		tpr = make([]float64, len(y)+1)
+		fpr = make([]float64, len(y)+1)
+	}
+	newEntry := true
+	var bin int = 1 // the initial bin is known to have 0 fpr and 0 tpr
+	var nPositive, nNegative float64
+	for i, u := range classes {
```

Comment:
I still think this can be better refactored. It has a long if statement that makes changes to newEntry which are non-permanent, because they are changed later.

It really feels like this can read something like
````
for i, u := range classes { 
     newBin = // check new bin
     if newBin() {
           bin++
           tpr[bin] = tpr[bin -1]
           fpr[bin] = fpr[bin-1]
     }
     newPos, newNeg := 1, 0
      if weights != nil {
           newPos = weights[i]
      }
      if !u {
          newPos, newNeg = newNeg, newPos
     }
     nPositive += newPos
     nNegative += newNeg
     tpr[bin] += newPos
    fpr[bin] += newNeg
}
````